File: Source/GI/spec_selftest.py
# Standalone verification of CascadeSpecification.hpp + the octahedral mapping.
# Mirrors the C++ math exactly so we can confirm the per-cascade table, the
# A4/B2 validators, VRAM, and the B1 octahedral round-trip without an MSVC build.
#
# Run:  python Source/GI/spec_selftest.py

import math, random

# ---- constants (mirror CascadeSpecification.hpp) ----
INTERVAL_LENGTH_FACTOR = 4.0

# ---- HierarchySpecification defaults ----
RC_QUERY_RADIUS   = 8.0     # m
BASE_INTERVAL_LEN = 0.15    # m  (t0)
BASE_PROBE_AXIS   = 64
BASE_ANGULAR_AXIS = 4       # R0

def resolve_interval_start(t0, c):
    return t0 * (INTERVAL_LENGTH_FACTOR ** c - 1.0) / 3.0

def resolve_interval_end(t0, c):
    return resolve_interval_start(t0, c + 1)

def resolve_base_probe_pitch():
    return BASE_INTERVAL_LEN

def resolve_cascade_count(override=0):
    if override != 0:
        return override
    diameter = 2.0 * RC_QUERY_RADIUS
    ratio    = 3.0 * diameter / BASE_INTERVAL_LEN + 1.0
    levels   = math.ceil(math.log(ratio) / math.log(INTERVAL_LENGTH_FACTOR))
    return max(1, int(levels))

def resolve_cascade(c):
    probe_axis = BASE_PROBE_AXIS >> c
    if probe_axis < 2:
        probe_axis = 2
    oct_side = BASE_ANGULAR_AXIS << c
    return {
        "c": c,
        "probe_axis": probe_axis,
        "oct_side": oct_side,
        "dirs": oct_side * oct_side,
        "int_start": resolve_interval_start(BASE_INTERVAL_LEN, c),
        "int_end":   resolve_interval_end(BASE_INTERVAL_LEN, c),
        "atlas_w": probe_axis * oct_side,
        "atlas_h": probe_axis * oct_side,
        "atlas_d": probe_axis,
    }

def resolve_hierarchy():
    return [resolve_cascade(c) for c in range(resolve_cascade_count())]

def confirm_angular_axis_pow2():
    a = BASE_ANGULAR_AXIS
    return a != 0 and (a & (a - 1)) == 0

def confirm_top_covers_region():
    count = resolve_cascade_count()
    top = resolve_cascade(count - 1)
    pitch = resolve_base_probe_pitch() * (2.0 ** (count - 1))
    return top["probe_axis"] * pitch >= 2.0 * RC_QUERY_RADIUS

def ray_budget(d):
    return d["probe_axis"] ** 3 * d["dirs"]

# ---- octahedral mapping, exact port of octahedral.glsl ----
def normalize3(v):
    l = math.sqrt(v[0]*v[0] + v[1]*v[1] + v[2]*v[2])
    return (v[0]/l, v[1]/l, v[2]/l)

def oct_encode(d):
    ax, ay, az = abs(d[0]), abs(d[1]), abs(d[2])
    denom = ax + ay + az
    ex, ey = d[0]/denom, d[2]/denom    # .xz
    if d[1] < 0.0:
        fx = (1.0 - abs(ey)) * (1.0 if ex >= 0.0 else -1.0)
        fy = (1.0 - abs(ex)) * (1.0 if ey >= 0.0 else -1.0)
        ex, ey = fx, fy
    return (ex * 0.5 + 0.5, ey * 0.5 + 0.5)

def oct_decode(e):
    cx, cy = e[0]*2.0 - 1.0, e[1]*2.0 - 1.0
    dx = cx
    dy = 1.0 - abs(cx) - abs(cy)
    dz = cy
    fold = max(-dy, 0.0)
    dx += -fold if dx >= 0.0 else fold
    dz += -fold if dz >= 0.0 else fold
    return normalize3((dx, dy, dz))

def main():
    print("=== HierarchySpecification ===")
    print(f"RcQueryRadius={RC_QUERY_RADIUS:.3f} m  t0={BASE_INTERVAL_LEN:.3f} m  "
          f"s0={resolve_base_probe_pitch():.3f} m  BaseProbeAxis={BASE_PROBE_AXIS}  R0={BASE_ANGULAR_AXIS}\n")

    r0pow2  = confirm_angular_axis_pow2()
    covered = confirm_top_covers_region()
    print(f"ConfirmAngularAxisPowerOfTwo  : {'PASS' if r0pow2 else 'FAIL'}")
    print(f"ConfirmTopCascadeCoversRegion : {'PASS' if covered else 'FAIL'}\n")

    count = resolve_cascade_count()
    print(f"ResolveCascadeCount = {count}\n")

    print(" c | ProbeAxis | OctSide |  IntStart |  IntEnd  |  Pitch  |    AtlasW x H x D    |    RayBudget   | VRAM(MB f16vec4)")
    print("---+-----------+---------+-----------+----------+---------+----------------------+----------------+-----------------")
    total_bytes = 0.0
    hier = resolve_hierarchy()
    for d in hier:
        pitch = resolve_base_probe_pitch() * (2.0 ** d["c"])
        b = d["atlas_w"] * d["atlas_h"] * d["atlas_d"] * 8.0   # vec4 f16 = 8 B/texel
        total_bytes += b
        print(f" {d['c']} |   {d['probe_axis']:5d}   |  {d['oct_side']:5d}  | "
              f" {d['int_start']:7.4f}  | {d['int_end']:7.4f}  | {pitch:6.4f}  | "
              f"{d['atlas_w']:5d} x {d['atlas_h']:5d} x {d['atlas_d']:4d} | "
              f"{ray_budget(d):14d} | {b/(1024*1024):8.2f}")

    print(f"\nTotal atlas VRAM (f16 vec4, single buffer) = {total_bytes/(1024*1024):.2f} MB")
    print(f"With a transient merge ping copy (worst)   = {2*total_bytes/(1024*1024):.2f} MB\n")

    top = hier[-1]
    tp = resolve_base_probe_pitch() * (2.0 ** top["c"])
    print(f"A4 top-cascade coverage: ProbeAxis({top['probe_axis']}) * pitch({tp:.3f}) = "
          f"{top['probe_axis']*tp:.3f} m  vs  diameter {2*RC_QUERY_RADIUS:.3f} m\n")

    # B1 round-trip
    rng = random.Random(1234567)
    max_err = 0.0; sum_err = 0.0; n = 0; outliers = 0
    for _ in range(10000):
        v = (rng.uniform(-1,1), rng.uniform(-1,1), rng.uniform(-1,1))
        if v[0]*v[0]+v[1]*v[1]+v[2]*v[2] < 1e-6:
            continue
        d = normalize3(v)
        r = oct_decode(oct_encode(d))
        e = math.sqrt((r[0]-d[0])**2 + (r[1]-d[1])**2 + (r[2]-d[2])**2)
        max_err = max(max_err, e); sum_err += e; n += 1
        if e > 1e-3 and outliers < 5:
            print(f"  outlier d={d} -> r={r} err={e:.5f}")
            outliers += 1
    print(f"B1 octahedral round-trip over {n} dirs: maxErr={max_err:.3e} "
          f"meanErr={sum_err/n:.3e}  {'PASS' if max_err < 1e-5 else 'FAIL'} (threshold 1e-5)")

    # texel-centre dirs unit length on c0 4x4
    max_off = 0.0
    for y in range(4):
        for x in range(4):
            dirv = oct_decode(((x+0.5)/4.0, (y+0.5)/4.0))
            ln = math.sqrt(dirv[0]**2 + dirv[1]**2 + dirv[2]**2)
            max_off = max(max_off, abs(ln - 1.0))
    print(f"Texel-centre dirs unit-length (c0 4x4): max |len-1| = {max_off:.3e} "
          f"{'PASS' if max_off < 1e-5 else 'FAIL'}")

if __name__ == "__main__":
    main()

File: Source/GI/spec_selftest.cpp
// Standalone self-test for CascadeSpecification.hpp + the octahedral mapping.
// Not part of the engine build; compiled ad hoc to verify the RC math core
// before any GPU pass depends on it. Prints the per-cascade table, runs the
// A4/B2 validators, sums VRAM, and runs the B1 octahedral round-trip (10k dirs).
//
// Build (from project root, inside a vcvars64 shell):
//   cl /std:c++17 /EHsc /O2 /nologo Source\GI\spec_selftest.cpp /Fe:spec_selftest.exe

#include "CascadeSpecification.hpp"

#include <cstdio>
#include <cmath>
#include <cstdint>
#include <random>

// ----- minimal vec3 + the octahedral mapping, CPU port of octahedral.glsl -----
struct V3 { float x, y, z; };
static V3   v3(float a, float b, float c) { return {a, b, c}; }
static float dot3(V3 a, V3 b) { return a.x*b.x + a.y*b.y + a.z*b.z; }
static V3   normalize3(V3 a) { float l = std::sqrt(dot3(a,a)); return {a.x/l, a.y/l, a.z/l}; }

struct V2 { float x, y; };

static V2 OctEncode(V3 Direction) {
    V3 AbsoluteDirection = { std::fabs(Direction.x), std::fabs(Direction.y), std::fabs(Direction.z) };
    float denom = AbsoluteDirection.x + AbsoluteDirection.y + AbsoluteDirection.z;
    V2 Encoded = { Direction.x / denom, Direction.z / denom };   // .xz
    if (Direction.y < 0.0f) {
        V2 Folded;
        Folded.x = (1.0f - std::fabs(Encoded.y)) * (Encoded.x >= 0.0f ? 1.0f : -1.0f);
        Folded.y = (1.0f - std::fabs(Encoded.x)) * (Encoded.y >= 0.0f ? 1.0f : -1.0f);
        Encoded = Folded;
    }
    return { Encoded.x * 0.5f + 0.5f, Encoded.y * 0.5f + 0.5f };
}

static V3 OctDecode(V2 Encoded) {
    V2 Centred = { Encoded.x * 2.0f - 1.0f, Encoded.y * 2.0f - 1.0f };
    V3 Direction = { Centred.x, 1.0f - std::fabs(Centred.x) - std::fabs(Centred.y), Centred.y };
    float Fold = std::fmax(-Direction.y, 0.0f);
    Direction.x += Direction.x >= 0.0f ? -Fold : Fold;
    Direction.z += Direction.z >= 0.0f ? -Fold : Fold;
    return normalize3(Direction);
}

int main() {
    using namespace Cascades;

    HierarchySpecification spec;   // defaults: R=8, t0=0.15, probes=64, R0=4

    std::printf("=== HierarchySpecification ===\n");
    std::printf("RcQueryRadius=%.3f m  t0=%.3f m  s0=%.3f m  BaseProbeAxis=%u  R0=%u\n\n",
                spec.RcQueryRadius, spec.BaseIntervalLength, ResolveBaseProbePitch(spec),
                spec.BaseProbeAxisCount, spec.BaseAngularAxisCount);

    const bool r0pow2  = ConfirmAngularAxisPowerOfTwo(spec);
    const bool covered = ConfirmTopCascadeCoversRegion(spec);
    std::printf("ConfirmAngularAxisPowerOfTwo  : %s\n", r0pow2  ? "PASS" : "FAIL");
    std::printf("ConfirmTopCascadeCoversRegion : %s\n\n", covered ? "PASS" : "FAIL");

    const uint32_t count = ResolveCascadeCount(spec);
    std::printf("ResolveCascadeCount = %u\n\n", count);

    std::printf(" c | ProbeAxis | OctSide |  IntStart |  IntEnd  |   Pitch  |   AtlasW x H x D    |   RayBudget   |  VRAM(MB f16vec4)\n");
    std::printf("---+-----------+---------+-----------+----------+----------+---------------------+---------------+------------------\n");

    double totalBytes = 0.0;
    auto hier = ResolveHierarchy(spec);
    for (const auto& d : hier) {
        const float pitch = ResolveBaseProbePitch(spec) * std::pow(2.0f, (float)d.CascadeIndex);
        const uint64_t budget = EstimateRayBudget(d);
        // payload = vec4 f16 (rgb + transmittance) = 8 bytes/texel
        const double bytes = (double)d.AtlasWidth * d.AtlasHeight * d.AtlasDepth * 8.0;
        totalBytes += bytes;
        std::printf(" %u |    %5u  |   %4u  |  %7.4f  | %7.4f  | %7.4f  | %5u x %5u x %4u | %13llu | %8.2f\n",
                    d.CascadeIndex, d.ProbeAxisCount, d.OctSide,
                    d.IntervalStart, d.IntervalEnd, pitch,
                    d.AtlasWidth, d.AtlasHeight, d.AtlasDepth,
                    (unsigned long long)budget, bytes / (1024.0*1024.0));
    }
    std::printf("\nTotal atlas VRAM (f16 vec4, single buffer) = %.2f MB\n", totalBytes / (1024.0*1024.0));
    std::printf("With a transient merge ping copy (worst case) = %.2f MB\n\n", 2.0 * totalBytes / (1024.0*1024.0));

    // Top-cascade coverage detail (A4)
    {
        const auto& top = hier.back();
        const float pitch = ResolveBaseProbePitch(spec) * std::pow(2.0f, (float)top.CascadeIndex);
        std::printf("A4 top-cascade coverage: ProbeAxis(%u) * pitch(%.3f) = %.3f m  vs  diameter %.3f m\n\n",
                    top.ProbeAxisCount, pitch, top.ProbeAxisCount * pitch, 2.0f * spec.RcQueryRadius);
    }

    // ----- B1: octahedral round-trip, 10k random unit dirs -----
    std::mt19937 rng(1234567u);
    std::uniform_real_distribution<float> uni(-1.0f, 1.0f);
    double maxErr = 0.0; double sumErr = 0.0; int n = 0; int worstReported = 0;
    for (int i = 0; i < 10000; ++i) {
        V3 d = { uni(rng), uni(rng), uni(rng) };
        float len2 = dot3(d, d);
        if (len2 < 1e-6f) continue;
        d = normalize3(d);
        V3 r = OctDecode(OctEncode(d));
        V3 diff = { r.x - d.x, r.y - d.y, r.z - d.z };
        double e = std::sqrt(dot3(diff, diff));
        maxErr = std::fmax(maxErr, e);
        sumErr += e; ++n;
        if (e > 1e-3 && worstReported < 5) {
            std::printf("  round-trip outlier: d=(%.3f,%.3f,%.3f) -> r=(%.3f,%.3f,%.3f) err=%.5f\n",
                        d.x,d.y,d.z, r.x,r.y,r.z, e);
            ++worstReported;
        }
    }
    std::printf("B1 octahedral round-trip over %d dirs: maxErr=%.3e meanErr=%.3e  %s (threshold 1e-5)\n",
                n, maxErr, sumErr / n, (maxErr < 1e-5) ? "PASS" : "FAIL");

    // Also check texel-centre directions are unit-length for c0 grid (4x4)
    {
        double maxOff = 0.0;
        // replicate DirectionFromOctTexel for OctSide=4
        for (int y = 0; y < 4; ++y)
            for (int x = 0; x < 4; ++x) {
                V2 e = { (x + 0.5f) / 4.0f, (y + 0.5f) / 4.0f };
                V3 dir = OctDecode(e);
                maxOff = std::fmax(maxOff, std::fabs(std::sqrt(dot3(dir,dir)) - 1.0));
            }
        std::printf("Texel-centre dirs unit-length (c0 4x4): max |len-1| = %.3e %s\n",
                    maxOff, (maxOff < 1e-5) ? "PASS" : "FAIL");
    }

    return (r0pow2 && covered && maxErr < 1e-5) ? 0 : 1;
}
